File: stacks/nearestGreaterElementOnRight.py
from stack import Stack

def nearestGreaterElementOnRight(arr):
    """using stack - O(n) implementation
    
    Arguments:
        arr {int} -- a list of numbers
    """
    s = Stack()
    greatArray = []
    for index, value in enumerate(arr):
        print(s.isEmpty())
        if index == len(arr) - 1:
            greatArray.append({value: -1})
            break
        
        while not s.isEmpty() and arr[s.top()] < value:
            print("in while")
            greatArray.append({arr[s.top()]: value})
            s.pop()
        else:
            print("in else")
            s.push(index)
    print(greatArray)
    return greatArray

nearestGreaterElementOnRight([5, 10, 7, 8, 3, 2, 12])

File: stacks/nearestGreaterElementOnRight.js
function nearestGreaterElementOnRight(arr){
  let greaterElements = []
  let stack = []
  for (let i = 0; i< arr.length; i++){
    console.log('******', stack)
    
    if (stack.length !== 0){
      while (arr[stack[stack.length - 1]] < arr[i]){
        greaterElements.push({[arr[stack.pop()]]: arr[i]})
        console.log("---pop---", stack)
      }
      // if (arr[stack[stack.length - 1]] > arr[i]){
        if (i === arr.length - 1){
          greaterElements.push({[arr[i]]: -1})
        } else{
          stack.push(i)
          console.log("+++push+++", stack)
        }
        
      // }
    } else {
      stack.push(i)
    }
  }
  console.log(greaterElements)
  return greaterElements
}

nearestGreaterElementOnRight([5, 10, 7, 8, 3, 2, 12])